cmd/vela-server/schedule.go

```diff
@@ -0,0 +1,384 @@
+		pipelineType = r.GetPipelineType()
+	)
+
+	// implement a loop to process asynchronous operations with a retry limit
+	//
+	// Some operations taken during this workflow can lead to race conditions failing to successfully process
+	// the request. This logic ensures we attempt our best efforts to handle these cases gracefully.
+	for i := 0; i < retryLimit; i++ {
+		logrus.Debugf("compilation loop - attempt %d", i+1)
+		// check if we're on the first iteration of the loop
+		if i > 0 {
+			// incrementally sleep in between retries
+			time.Sleep(time.Duration(i) * time.Second)
+		}
+
+		// send API call to attempt to capture the pipeline
+		pipeline, err = database.GetPipelineForRepo(b.GetCommit(), r)
+		if err != nil { // assume the pipeline doesn't exist in the database yet
+			// send API call to capture the pipeline configuration file
+			config, err = scm.ConfigBackoff(u, r, b.GetCommit())
+			if err != nil {
+				return fmt.Errorf("unable to get pipeline config for %s/%s: %w", r.GetFullName(), b.GetCommit(), err)
+			}
+		} else {
+			config = pipeline.GetData()
+		}
+
+		// send API call to capture repo for the counter (grabbing repo again to ensure counter is correct)
+		r, err = database.GetRepoForOrg(r.GetOrg(), r.GetName())
+		if err != nil {
+			err = fmt.Errorf("unable to get repo %s: %w", r.GetFullName(), err)
+
+			// check if the retry limit has been exceeded
+			if i < retryLimit-1 {
+				logrus.WithError(err).Warningf("retrying #%d", i+1)
+
+				// continue to the next iteration of the loop
+				continue
+			}
+
+			return err
+		}
+
+		// update repo fields with any changes from SCM process
+		r.SetTopics(r.GetTopics())
+		r.SetBranch(r.GetBranch())
+
+		// set the build numbers based off repo counter
+		r.SetCounter(r.GetCounter() + 1)
+		b.SetNumber(r.GetCounter() + 1)
+		// set the parent equal to the current repo counter
+		b.SetParent(r.GetCounter())
+		// check if the parent is set to 0
+		if b.GetParent() == 0 {
+			// parent should be "1" if it's the first build ran
+			b.SetParent(1)
+		}
+
+		// set the build link if a web address is provided
+		if len(metadata.Vela.WebAddress) > 0 {
+			b.SetLink(fmt.Sprintf("%s/%s/%d", metadata.Vela.WebAddress, r.GetFullName(), b.GetNumber()))
+		}
+
+		// ensure we use the expected pipeline type when compiling
+		//
+		// The pipeline type for a repo can change at any time which can break compiling
+		// existing pipelines in the system for that repo. To account for this, we update
+		// the repo pipeline type to match what was defined for the existing pipeline
+		// before compiling. After we're done compiling, we reset the pipeline type.
+		if len(pipeline.GetType()) > 0 {
+			r.SetPipelineType(pipeline.GetType())
+		}
+
+		var compiled *library.Pipeline
+		// parse and compile the pipeline configuration file
+		p, compiled, err = compiler.
+			Duplicate().
+			WithBuild(b).
+			WithMetadata(metadata).
+			WithRepo(r).
+			WithUser(u).
+			Compile(config)
```

Comment:
```suggestion
		p, compiled, err = compiler.
			Duplicate().
			WithBuild(b).
			WithCommit(commit).
			WithMetadata(metadata).
			WithRepo(r).
			WithUser(u).
			Compile(config)
```
This is new 😅 ([ref](https://github.com/go-vela/server/pull/859)). FWIW, adding this line doesn't actually change anything because an empty commit [here](https://github.com/go-vela/server/blob/ffac676b781bab77d57caccbc12edd33d26e66b4/compiler/native/expand.go#L238) just uses the default branch. But I suppose it's nice to not rely on that being true and add the commit ourselves 🤷‍♂️ .